### PR TITLE
[DependencyInjection] Add the Required attribute

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | TODO

This PR proposes a new attribute `#[Required]` that can be used instead of the `@required` annotation.

### CODE REVIEW

File: src/Symfony/Contracts/Service/Attribute/Required.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\Service\Attribute;
+
+use Attribute;
+
+/**
+ * Flags a public setter, wither or property for autowiring.
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/Service/Attribute/Required.php:17]: Wrong :)
The difference is stubble yet critical: this is *not* for autowiring.
This is to improve the expressivity of the source code, to tell that this property or method should be considered as part of the instantiation steps of this class.
Then, it happens that this info is useful to autowiring.
But let's not short-circuit the meaning of things please :)

Comment (derrabus) [reply src/Symfony/Contracts/Service/Attribute/Required.php:17]: You are right! I'll try to find a better description.

Comment (derrabus) [reply src/Symfony/Contracts/Service/Attribute/Required.php]: Please have a look if the new description is a better fit.

**Comment by derrabus** [comment]
The PR has been updated for php 8 syntax changes and is now compatible with php 8.0.0-beta1.

**Comment by derrabus** [comment]
The PR has once again been updated because of a syntax change: https://github.com/php/php-src/commit/8b37c1e993f17345312a76a239b77ae137d3f8e8

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php [hunk 2]
```diff
@@ -49,6 +50,14 @@ protected function processValue($value, bool $isRoot = false)
             }
 
             while (true) {
+                if (method_exists($r, 'getAttributes') && $r->getAttributes(Required::class)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:53]: let's use PHP_VERSION_ID, that will ease spot the line when PHP8 will be the minimum supported version (yes, that will happen one day :) )

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php [hunk 3]
```diff
@@ -49,6 +50,14 @@ protected function processValue($value, bool $isRoot = false)
...
+                    if ($this->isWither($r, $r->getDocComment() ?: '')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:54]: btw, shouldn't isWither() return `true` when the `static` return type is used?

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php [hunk 4]
```diff
@@ -45,10 +46,9 @@ protected function processValue($value, bool $isRoot = false)
             if (!($type = $reflectionProperty->getType()) instanceof \ReflectionNamedType) {
                 continue;
             }
-            if (false === $doc = $reflectionProperty->getDocComment()) {
-                continue;
-            }
-            if (false === stripos($doc, '@required') || !preg_match('#(?:^/\*\*|\n\s*+\*)\s*+@required(?:\s|\*/$)#i', $doc)) {
+            if ((!method_exists($reflectionProperty, 'getAttributes') || !$reflectionProperty->getAttributes(Required::class))
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredPropertiesPass.php:49]: PHP_VERSION_ID

**Review by nicolas-grekas — COMMENTED** [review]
thanks, looks nice, here are some minor comments
btw, if you want to have a look at making the PHP8 build green (or just have less failures), that'd be awesome :)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php (hunk 3 above)
Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowireRequiredMethodsPass.php:54]: Yes, that's kind of the point of that method (#36636).

**Comment by nicolas-grekas** [comment]
The failure on travis can be skipped by listing the fixture file in `.github/patch-types.php` maybe?

**Comment by derrabus** [comment]
I've added the file there. 👍

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @derrabus.

**Comment by TomasVotruba** [comment]
This change is now covered in @rectorphp https://github.com/rectorphp/rector/pull/4348/files#diff-8e9faa1ff0eb70ba27a9918bf9b062b4

Let me know if we got it right :wink:

**Comment by derrabus** [comment]
@TomasVotruba Your fixture looks correct. In addition to the case that it covers, the `@required` annotation could also be attached to a public typed property.

**Comment by TomasVotruba** [comment]
@derrabus Thanks for feedback :+1: 

It was not clear from the tests, but property is supported as well.
I've added the test fixture to have it covered explicitly too:
https://github.com/rectorphp/rector/pull/4353/files

**Pull request merged**